Implement --result-diff option.

cheqianh opened the pull request:
### Description: 
Working on new extension (--result-diff, -R) which is able to compare two implementations' behaviors provided by a result file. This PR links #3.
**Usage**: `ion-test-driver --results-diff imp1_master imp1_commit result_file.ion -o output`
**Note** that the order of --results-diff matters, our logic compares imp1_commit against to imp1_master, and the comparison report will show changes based on imp1_master.

### Logic
First of all, assuming that cli process and compare features are correct. If there is any issue caused by cli tool, then the analyze will stop.

An example result looks like below:
```
'file2': {
	'ion-java': {result: PASS}, 
	'ion-c': {
			result: FAIL,
                        read_error: ErrorReport::[{
                                error_type: READ,
                                message: "some thing happened....",
                                location: "file2.ion"
                          }],
                        read_compare: {
                                errors: [],
                                failures: ComparisonReport::[{
                                      result: NOT_EQUAL,
                                      lhs: {
                                         location: "ion-c_abcd123.ion",
                                          event: {
                                              event_type: SCALAR,
                                              ion_type: INT,
                                              ...
                                          },
                                          event_index: 2
                                      },
                                      rhs: {
                                          location: "test_file_4.ion",
                                          event: {
                                              event_type: SCALAR,
                                              ion_type: INT,
                                              ...
                                           },
                                           event_index:2
                                      },
                                     message: "1 vs. 2"
                                }]
                         }
			write_error: ErrorReport::[
          		{
            			error_type: WRITE,
            			message: "some weird reason....",
            			location: "some location...",
            			event_index: 11
          		}],	
		}
	}
},
```
I'm going to parse the error report into some data structures (read_error, read_compare, write_error, write_compare) first and then compare them after parsing. 

When we analysis the result, there are four phases: read_error, read_compare, write_error, write_compare. We need to compare two revisions' results and analysis them in the order below:
1. Check `results` field, if both of them are PASS then go to next test file. otherwise, go to next step (still same test file).
2. Check `read_error` field. If both of them have same read_error or don't have any error, go to next step. Otherwise, write `read performance changed` error to the final result and go to next step.
3. Check `read_compare` field. We analyze the given read_compare report and find all disagree pairs. For example, there are four disagree reports: `impl_1_master` disagrees with `imp_2` and `imp_2` disagrees with `impl_1_master` as well. Also `impl_1_commit` disagrees with `impl_1_master` and  `impl_1_master` disagrees with `impl_1_commit` as well. but `impl_1_commit` and `impl_1_master` are both agree with `imp_2`. So we can transfer this to some data struct (call it disagree list) for two revisions that we compare: 
`impl_1_master_disagree_list`: `impl_2`, `impl_1_commit` and
`impl_1_commit_disagree_list`: `impl_1_master` 
we also create a mark to describe if two revisions agree with each other. For the case above, they disagree with each other. After extracting two disagree lists, we need compare master branch and new commit by below two cases:
**3.1.** If they agree with each other, their disagree lists should be same. Raise a `cli compare diff` error if they are not same.
**3.2.** If they disagree with each other, we will write down what implementations that the master commit are no longer agree with and what implementation the master commit start agreeing with. We need to write the comparison report to describe why it disagrees some implementations as well.
**Note** that we only care about the comparison report involving two implementations that we compare.
4. Check `write_error`, same as step 2.
5. Check `write_compare`, same as step 3.

After comparing two implementations, write the report to the final result.

### Example
To test two implementations ion-c-master and ion-c-commit for file2
```
'file2': {
	'ion-c-master': {
                          result: FAIL,
                          read_compare: {
                                errors: [],
                                failures: ComparisonReport::[
                                  {
                                      result: NOT_EQUAL,
                                      lhs: {
                                         location: "ion-c_master.ion",
                                         ......
                                       },
                                       rhs: {
                                           location: "file2.ion",
                                           ......
                                       },
                                       message: "1 vs. 2"
                                 },
                                 {
                                      result: NOT_EQUAL,
                                      lhs: {
                                         location: "ion-c_commit.ion",
                                         ......
                                       },
                                       rhs: {
                                           location: "file2.ion",
                                           ......
                                       },
                                       message: "1 vs. 2"
                                 },
                                 ......
                             ]
                          }
                   }, 
	'ion-c-commit': {
			result: FAIL,
                        read_error: ErrorReport::[{
                                error_type: READ,
                               message: "some reason caused by new commit.....",
                                location: "file2.ion"
                          }],
                        read_compare: {
                                errors: [],
                                failures: ComparisonReport::[
                                    {
                                      result: NOT_EQUAL,
                                      lhs: {
                                         location: "ion-c_master.ion",
                                         ......
                                       },
                                       rhs: {
                                           location: "file2.ion",
                                           ......
                                       },
                                       message: "1 vs. 2"
                                  },
                                   {
                                      result: NOT_EQUAL,
                                      lhs: {
                                         location: "ion-c_commit.ion",
                                          ......
                                      },
                                      rhs: {
                                          location: "file2.ion",
                                          ......
                                      },
                                     message: "1 vs. 2"
                                  },
                                  .......
                              ],
                         }
		}
	}
},
```
We first check `result` field, since they both 'PASS' so we go next step.
Next we check `read_error`, ion-c-commit has a read_error while ion-c-master performance correctly so we write `read performance changed` to report.
Next we check `read_compare`, we first check if two read_compare are same and they are same. Next we extract disagree list from the report. After that, we found that they all disagree to the original file2.ion but agree with each other. So the disagree list for them are:
ion-c-master: `file2`
ion-c-commit: `file2`
Since they agree with each other and there is no new disagree implementations need to be written to the final result, the final result will only have one report: ion-c-commit changed the read performance:
```
{
   bad: {
        file2: {
            read_error: {
                message: Read performance is changed in ion-c-commit.
                original_read_error: ErrorReport::[{
                                error_type: READ,
                                message: "some reason caused by new commit.....",
                                location: "file2.ion"
                 }],
            },
            read_compare: {
                message: "Nothing changed",
                new_disagree_impl: [],
                new_agree_impl: [],
                original_read_compare: read_compare: {
                               errors: [],
                               failures: ComparisonReport::[ ..... ]
               }
            },
            write_error: {},
            write_compare: {}
        },
       ....
   },
   good: {
      ....
   },
   ......
}
```


### Implementation plan:
There are 5 things need to implement to finish the project. (it matches TODO # in the code below):
- [x] **TODO0** Create structures to describe each compare result (read_error, read_compare, write_error, write_compare).
- [x] **TODO1** Decide the spec of the description (since the revision can be either hash or branch name. So need to write some logic to parse the description and figure out how to match branch name -> hash).
- [x] **TODO2** Parse file content to structures.
- [x] **TODO3** Analyze two implementations' results.
- [x] **TODO4** write compare result to final report. Also need to support -o options.

### Test

- [x] **TODO 6** Working on the unit tests.

cheqianh committed:
Working on new extension. (implementation plan)

tgregg commented on amazon/iontest/ion_test_driver.py:
```diff
@@ -18,6 +18,7 @@
     ion_test_driver.py [--implementation <description>]... [--ion-tests <description>] [--test <type>]...
                        [--local-only] [--cmake <path>] [--git <path>] [--maven <path>] [--java <path>] [--npm <path>]
                        [--node <path>] [--output-dir <dir>] [--results-file <file>] [<test_file>]...
+    ion_test_driver.py [--results-diff <description>]... [<result_file>]
```
Maybe it should just be
`ion_test_driver.py [--results-diff <description1> <description2>] [<result_file>]`

So you only need to specify `--results-diff` once and it must always accept exactly two descriptions?

cheqianh replied:
line 21

tgregg commented on amazon/iontest/ion_test_driver.py:
```diff
@@ -55,6 +56,9 @@
                                         `--output-dir` option.
 
+    -R, --results-diff <description>    Given two descriptions. analysis the result files and shows
+                                        the difference between these two implementations.
```
I suggest:

"Given two implementation descriptions, analyze an existing results file to identify any differences between the two implementations."

cheqianh replied:
line 59

tgregg commented on amazon/iontest/ion_test_driver.py:
```diff
@@ -741,13 +748,96 @@ def parse_implementations(descriptions, output_root):
+
+def analyze_results(implementations, results_file):
+    with open(results_file, 'r') as f:
+        result_context = f.read()
```
I would recommend just using `simpleion.load` to load the entire results file into a Python object. That should simplify this logic a lot.

cheqianh replied:
line 756

tgregg commented on amazon/iontest/ion_test_driver.py:
```diff
@@ -741,13 +748,96 @@ def parse_implementations(descriptions, output_root):
+    with open(results_file, 'r') as f:
+        result_context = f.read()
+
+    # TODO 1. decide how to represent implementations' description, currently is string (e.g "ion-java_abcd123")
```
You can probably reuse logic like [`parse_implementations`](https://github.com/cheqianh/ion-test-driver/blob/master/amazon/iontest/ion_test_driver.py#L739) and [`IonImplementation`](https://github.com/cheqianh/ion-test-driver/blob/master/amazon/iontest/ion_test_driver.py#L176) from `ion_test_driver.py`.

tgregg reviewed (commented):
Quick note: for clarity, please use the term "behavior" rather than "performance" in the cases you're describing. Typically "performance" is used to describe the runtime characteristics of a program, like execution time, memory consumption, etc.

I think overall your plan sounds good. I have a couple of suggestions below.

cheqianh committed:
Fix above issues and TODO4

cheqianh commented:
This commit fixed issues above and implement TODO 0, 1, 2, 4. 

TODO 0 and 2 will be easy to implement After using dump/load.
For TODO1 the description's format is `name,revision` now since the location(URL) is unnecessary for analyzing the results file. 
For TODO4 it supports -o options now.

tgregg commented on amazon/iontest/ion_test_driver.py:
```diff
@@ -55,6 +56,10 @@
                                         `ion-test-driver-results.ion` under the directory specified by the
                                         `--output-dir` option.
 
+    -R, --results-diff <description1>   Given two implementation descriptions of the form name,revision. Name is the
+    <description2>                      implementation's name and revision should be commit hash. Analyze an existing
```
I don't think this format will work with docopt. You'll probably need to have `-R, --results-diff <description1> <description2>` all on the same line.

tgregg commented on amazon/iontest/ion_test_driver.py:
```diff
@@ -741,13 +746,58 @@ def parse_implementations(descriptions, output_root):
+        output_root = os.path.abspath(arguments['--output-dir'])
+        if arguments['--output-dir'] is '.':
+            output_root = os.path.join(output_root, 'result.ion')
+        first_implementation = arguments['--results-diff']
+        second_implementation = arguments['<description2>']
```
This doesn't look quite right. You'll probably have to test it out.

tgregg reviewed (commented):
The framework you've set up looks good.

cheqianh committed:
Able to analyze read_error,read_compare,write_error.

cheqianh commented:
New commit f7fd868
Fixed issues above, and worked on TODO 3.
Currently able to analyze `result`, `read_error`, `read_compare` and `write_error`. `Write_compare` will be finished soon.

cheqianh committed:
Able to analyze write_compare.

Change term performance to behavior.

Fix same implementation restrict issue.

Added original file check and fix filename for write_compare.

cheqianh commented:
commit d9031ac Finished write_compare and refactor some functions.

cheqianh commented:
### Here are example analyze result(I modified the file to cause errors for each step on purpose):
**1. check result field:**
they should both pass or fail, otherwise it may be a cli issue.
![Screen Shot 2020-10-14 at 12 40 49 PM](https://user-images.githubusercontent.com/67451029/96037301-81875280-0e1a-11eb-9a97-72d4c203dce1.png)

**2. read_error:**
they should have the same errors, but one of the revisions has extra message field: `meessage: "test"`
![Screen Shot 2020-10-14 at 12 41 57 PM](https://user-images.githubusercontent.com/67451029/96037414-aa0f4c80-0e1a-11eb-9ec4-66e29c0bb2d8.png)


**3. read_compare**
2 revisions of ion-java disagree with each other but both of them disagree with ion-c. So it's a behavior change rather than error.
![Screen Shot 2020-10-14 at 12 42 31 PM](https://user-images.githubusercontent.com/67451029/96037474-bdbab300-0e1a-11eb-839e-6c89832b2dcc.png)

**4. write_error:**
The example below have different errors, it may be the write behavior changed. the reason of the example below is ion-java_16cb1460 revision fixed the write error in ion-java_bed94a9b so it doesn't show up again. the author of the new commit should look into that and decide if he wants to merge.
![Screen Shot 2020-10-14 at 12 50 36 PM](https://user-images.githubusercontent.com/67451029/96038370-e0999700-0e1b-11eb-9ecb-3549a085558d.png)

**5:write_compare:**
write_compare phase compares total 2x^2 + 1 files where x is the number of implementations. So for each implementations, they have multiple disagree lists corresponding different situation.
For example below, the first implementation `ion-java_bed94a9b`, it's first disagree list's key is `'binary,ion-java_bed94a9b'` which means using `ion-java_bed94a9b` to rewrite `ion-java_bed94a9b` that generate in read phase in binary format. and the elements in the lists are files that it disagrees.
![Screen Shot 2020-10-14 at 12 54 00 PM](https://user-images.githubusercontent.com/67451029/96038684-5aca1b80-0e1c-11eb-881a-80c7e2a752e3.png)

cheqianh commented:
commit 2676296 Finished write_compare and refactor some functions.

I used example file that generates errors/failures of each steps to test each functions. Since we need to look into the result file manually to decide if they work as expected, Is it enough for testing? 

### Here are example analyze result(I modified the file to cause errors for each step on purpose):
**1. check result field:**
I made one revision PASS and the other one FAIL. Since they should both pass or fail, so it raise an error.
![Screen Shot 2020-10-14 at 12 40 49 PM](https://user-images.githubusercontent.com/67451029/96037301-81875280-0e1a-11eb-9a97-72d4c203dce1.png)

**2. read_error:**
they should have the same errors, but one of the revisions has extra message field: `meessage: "test"` so it might be read behavior changed.
![Screen Shot 2020-10-14 at 12 41 57 PM](https://user-images.githubusercontent.com/67451029/96037414-aa0f4c80-0e1a-11eb-9ec4-66e29c0bb2d8.png)


**3. read_compare**
2 revisions of ion-java disagree with each other but both of them disagree with ion-c. So it's a behavior change rather than error.
![Screen Shot 2020-10-14 at 1 05 26 PM](https://user-images.githubusercontent.com/67451029/96039714-f4de9380-0e1d-11eb-8c36-51f790426916.png)


**4. write_error:**
The example below have different errors, it may be write behavior changed. the reason of the example below is ion-java_16cb1460 revision fixed the write error in ion-java_bed94a9b so it doesn't show up again. the author of the new commit should look into that and decide if he wants to merge.
![Screen Shot 2020-10-14 at 12 50 36 PM](https://user-images.githubusercontent.com/67451029/96038370-e0999700-0e1b-11eb-9ecb-3549a085558d.png)

**5:write_compare:**
write_compare phase compares total 2x^2 + 1 files where x is the number of implementations. So for each implementations, they have multiple disagree lists corresponding different situations.
For example below (highlight struct), the first implementation `ion-java_bed94a9b`, it's first disagree list's key is `'ion-java_bed94a9,binary,ion-java_bed94a9b'` which means the file that uses `ion-java_bed94a9b` implementation to rewrite `ion-java_bed94a9b` file in binary format. and the elements in the lists are files that it disagrees.
![Screen Shot 2020-10-14 at 2 13 42 PM](https://user-images.githubusercontent.com/67451029/96045903-7edf2a00-0e27-11eb-9c79-39a572ea2ad9.png)

cheqianh committed:
Improve text.

tgregg commented:
I think the logic is complicated enough that this change should include unit tests. The manual testing you've done can serve as the basis for the unit tests: commit data files that have been manually changed so that they are expected to produce a particular result. Run the results diff on each file and verify that it produces the expected Ion output.

tgregg commented:
For your example "4. write_error" above, isn't this particular example a case where we don't need to alert the user? The previous implementation caused an unexpected error, while the new one does not.

tgregg commented on amazon/iontest/ion_test_driver.py:
```diff
@@ -741,13 +747,348 @@ def parse_implementations(descriptions, output_root):
+                                                  "file is the file that be re-writen",
+                    "disagree_lists": {first_impl: first_disagree_list_for_write,
+                                       second_impl: second_disagree_list_for_write},
+                    "no_more_agree_with": no_more_agree_lists,
+                    "start_agree_with": start_agree_lists,
```
I suggest "no_longer_agrees_with" and "now_agrees_with".

cheqianh committed:
Improve the read_error/write_error report and rename some variables.

cheqianh commented:
Thanks for the cr, new commit 241468b:

1. Changed the logic for read_error and write_error, Now it will only be written to report if it raise a new error that wasn't found in previous commit.
For example, the example output above for step four now don't write anything to report.
If I switch the write_error field, which means the old commit doesn't have any error and the new commit has. The report is:
![Screen Shot 2020-10-14 at 5 09 17 PM](https://user-images.githubusercontent.com/67451029/96058430-03d63d80-0e40-11eb-9a7a-f1860f3a7de9.png)



2. Created variables for each fields such as `result`, `no_longer_agrees_with`, `now_agrees_with` etc.

3. I will create unit tests in the next commit.

cheqianh commented:
In case I didn't understand it correctly, the term `'unit'` for unit test here represents a test file that will cause ion-test-driver to alert users rather than a function within ion-test-driver.py right?

I've already created some test files and expect outputs, I wrote another small python project to run the ion-test-driver for each file and compare them to the expect files. Do I have to use a python unit test framework?

tgregg commented:
Yes, a [unit test](https://en.wikipedia.org/wiki/Unit_testing).

Like the tests we have in the [tests](https://github.com/amzn/ion-python/tree/master/tests) directory of ion-python. You should be able to use the test files and expected outputs you've already generated, and just write python code to verify them against the behavior of the ion-test-driver in an automated way using pytest. That way, whenever ion-test-driver is built, we can verify that it still behaves as expected without having to use some external tool or run commands by hand and manually verify output.

cheqianh committed:
Added unit test and able to support two types of description.

cheqianh commented:
Got it, thanks. 

**For new commit 9af84941 :**
1. Added unit test, tested each branch that shows up in -R logic. below two new commits fix some style issues.
2. Update pytest to 6.1.1.
3. Description for -R option can be either `name,commit_hash` or `name,location,revision` now in case users don't know the commit hash.

**For unit test:**
All tests are located in `tests` file. Main test logics are in `test_res_diff.py`. file `res_diff_tests` includes test files and `res_diff_expect` includes expect output.

**File structure:**
![Screen Shot 2020-10-16 at 1 49 37 PM](https://user-images.githubusercontent.com/67451029/96307431-76692980-0fb6-11eb-816b-b6129c907204.png)


**Example pytest run:**
![Screen Shot 2020-10-16 at 11 22 25 AM](https://user-images.githubusercontent.com/67451029/96294873-e0c39f00-0fa1-11eb-88db-0c56d1be6f73.png)

cheqianh committed:
Styles.

cheqianh committed:
Change style issues.

tgregg commented on tests/util.py:
```diff
@@ -0,0 +1,37 @@
+
+
+def compare_two_files(first_file, second_file):
+    return filecmp.cmp(first_file, second_file, shallow=False)
```
Ion files should always be compared for Ion data model equivalence rather than strict content equivalence. Use the `equivalence` module from ion-python.

tgregg commented on tests/util.py:
```diff
@@ -0,0 +1,37 @@
+def run_ion_test_driver(test_file, first_desc, second_desc):
+    dir_name, file = os.path.split(test_file)
+    output_name = os.path.join(dir_name, 'test_' + file)
+    _, stderr = Popen(('python3', ION_TEST_DRIVER_PATH, '-R', first_desc, second_desc, test_file, '-o', output_name),
```
Should there be a failure if anything is written to `stderr`?

Also, the command invocation should either return 0 (success) or non-zero (failure). Github Actions should automatically fail if one of its scripts returns non-zero. These unit tests should verify that the command returns success when expected to be successful, and failure when expected to fail.

cheqianh replied:
Yeah, I changed this function so we can get the return value first and check if the return value is expected.

tgregg commented on tests/res_diff_expect/test_diff_results.ion:
```diff
@@ -0,0 +1,8 @@
+{
+ test: {
+  result: {
+   message: "Result: two revisions have different results. This might be a cli tool issue."
```
I'd drop the `This might be a cli tool issue` guidance, because it could just as easily be a bug in the actual ion-java implementation.

cheqianh committed:
ion-python version to 6

tgregg commented on tests/res_diff_tests/test_fix_read_compare_errors.ion:
```diff
@@ -0,0 +1,22 @@
+             }
+        },
+        'ion-java_2': {
+            result:FAIL
```
Please help me remember... do `bad` tests return a `FAIL` result when parsing fails (which is suspected)? Or do they return `PASS` when parsing fails?

If the former, shouldn't this difference cause a `--results-diff` to return a failure, since the error reports are different?

Sorry if this causes churn, but I think we should actually print results from `--results-diff` whenever there are changes between the two revisions (good or bad), but only return non-zero (i.e. failure) from the command if the differences amount to a regression or require examination from the developer to determine whether they are an improvement.

cheqianh replied:
If the result field passes or not is depending on the other four phases (read, read_compare, write, write_compare). It's FAIL as long as we write anything to report for these four phases. So the pass/fail depends on if the implementations behavior same rather than if they face errors.

So even though it faces a parsing failure, it's still treated as PASS as long as all implementations has the same incomplete event stream output.

Sure, Now we check two things: return value and output. return value should be a expected value (0 for pass and non-zero for fail) and output should be same as expected output.

tgregg commented on tests/res_diff_expect/test_result_both_fail.ion:
```diff
@@ -0,0 +1,20 @@
+ },
+ test2: {
+  result: {
+   message: "Result: two revisions have different results. This might be a cli tool issue."
```
Similar to one of my previous questions, can you remind me if for `bad`, going from `FAIL` to `PASS` is good or bad?

I think either way, we should show here that the change occurred, but the `--results-diff` invocation should only return a failure code (thereby failing the Github Actions run) if the change was for the worse.

tgregg commented on tests/res_diff_expect/test_result_both_fail.ion:
```diff
@@ -0,0 +1,20 @@
+{
+ test: {
+  read_error: {
+   message: "Read_error: new commit causes new read error(s).",
```
I'm thinking we should just show the complete change. Maybe change the message to `Read_error: new commit has different read errors(s).` and include the errors from both revisions here. That will make it easier for the user to evaluate the difference.

cheqianh replied:
Added. We now write all difference into the report but only alert user once the new commit has errors that the old didn't have.

tgregg commented on tests/res_diff_expect/test_revisions_with_diff_list.ion:
```diff
@@ -0,0 +1,17 @@
+{
+ test: {
+  read_compare: {
+   message: "Read_compare: two revisions agree with each other but have different disagree lists. This might be a cli tool issue.",
```
Again, I would remove `This might be a cli tool issue.` from everywhere. It could be a problem in the CLI or the implementation itself; either way, it should be fixed.

In this case, since the implementations that the latter revision disagrees with is a subset of the implementations that the former revision disagrees with, should we print this difference but not fail the `--results-diff` invocation due to this?

cheqianh replied:
I think two revisions may be supposed to have the same disagree list as long as they agree with each other, right?

For example, if impl_1 agrees with impl_2 and impl_2 disagrees with impl_3. impl_1 has to disagree with impl_3, right? Otherwise It's an issue.

tgregg commented on tests/res_diff_expect/test_write_disagree.ion:
```diff
@@ -0,0 +1,44 @@
+{
+ test: {
+  write_compare: {
+   message: "Write_compare: write behavior changed. Description below following the format: 'impl,type,file' where impl is the implementation used for writing files,type is either 'text' or 'binary' and file is the file that be re-writen",
```
"re-writen" -> "re-written"

I think it would be good to also mention what the file name means. It's the implementation that generated the write instruction from which the current implementation is re-writing, right?

cheqianh replied:
Sure, it's the implementation that is using the 'process' command. I added comment to the message field. (ion_test_driver.py 1107)

tgregg commented on tests/test_res_diff.py:
```diff
@@ -0,0 +1,238 @@
+EXPECT_FILE_PATH = os.path.join(os.path.split(os.path.abspath(__file__))[0], EXPECT_FILE_NAME)
+
+
+# Empty report.
+def test_empty_report():
+    test_file = 'test_empty_report.ion'
+    output_path = run_ion_test_driver(os.path.join(TEST_FILE_PATH, test_file), 'ion-java,1', 'ion-java,2')
+    res = compare_two_files(output_path, os.path.join(EXPECT_FILE_PATH, test_file))
+    os.remove(output_path)
+    assert res is True
```
All of these tests have exactly the same structure and differ only by the test file, making them an excellent use case for a reusable function or (even better) [parameterized tests](https://docs.pytest.org/en/stable/example/parametrize.html).

ion-python has parameterized tests (e.g. [here](https://github.com/amzn/ion-python/blob/master/tests/test_vectors.py#L360)), but it uses some [custom parameterization logic](https://github.com/amzn/ion-python/blob/b6ce1defa4973825623cc95ac4be3203bbe1f69d/tests/__init__.py#L46) that you might be able to get working natively using pytest.

cheqianh replied:
Got it, created parameterized tests for each cases.

tgregg reviewed (commented):
The diff includes some `.DS_Store` files that should not be committed.

The test data is good, but I recommend a few additions (I apologize if I missed any of these):
* For the test data I saw that had differences between revisions, the differences were always in the same phase, e.g. read_compare vs read_compare. Can you also test what happens when one implementation has a failure in e.g. read_compare while the other has a failure in write_compare?
* Does there need to be testing for equivs/non-equivs results too?

cheqianh committed:
Added more test cases and fix some issues.

cheqianh commented:
Besides comments above, here is a summary list of what I did for new commit:
1. Now we can get both return value and output to decide if it functions as expected. we assert two things: 1. if the return value is 0 or non-zero and if it's what we expect. 2. If the analysis result output same as the expect output. (ion-test-driver.py 913-914, 1125)

2. Now we will write all the differences to report but only alert users when raising new errors.
(ion-test-driver.py 971, 998, 1055, 1079)

3. Improved the message field such as removing `"this is a cli issue"`, adding field description for write_compare phase and so on. (ion-test-driver.py 1107)

4. Renamed test files. Every files have a prefix which can be either` 'pass'` or `'fail'`. pass means that ion-test-driver will return a zero for this file and fail means that ion-test-driver will return a non-zero value for it.
For example below, this file will return a non-zero value and fail the github actions.
<img width="113" alt="Screen Shot 2020-10-19 at 4 56 41 PM" src="https://user-images.githubusercontent.com/67451029/96524055-124f9b00-122c-11eb-81bd-e87feac4ed41.png">


5. Added more test cases:
If two revisions have errors in different phases, the ion-test-driver should write the first error it detects into the report and go the next file.
**pass_empty_report**: an empty report.
**fail_read_errors_vs_other_errors**: old revision has read errors (phase one) but the new one has later phase errors.
**fail_read_compare_vs_other_errors_1**: old revision has read compare (phase two) errors but the new one has early phase errors.
**fail_read_compare_vs_other_errors_2**: old revision has read compare (phase two) errors but the new one has later phase errors.
**fail_write_errors_vs_other_errors_1**: old revision has write errors (phase three) errors but the new one has early phase errors.
**fail_write_errors_vs_other_errors_2**: old revision has write errors (phase three) errors but the new one has later phase errors.
**fail_write_compare_errors_vs_other_errors:** old revision has write compare errors (phase four) errors but the new one has early phase errors.
&nbsp;
For Equivs/non-equivs it writes errors to read_compare/write_compare phase, they are same as read_compare_failures but with different messages. I added test files below.
**fail_equivs_failures.ion**: when new commit writes equivs failures to report.
**fail_equivs_missing_failures.ion**: when new commit missing equivs failures but the old one has.
**fail_non_equivs_failures.ion**: when new commit writes non-equivs failures to report.
**fail_non_equivs_missing_failures.ion**: when new commit missing non-equivs failures but the old one has.

Current test output: 
![Screen Shot 2020-10-20 at 3 24 08 PM](https://user-images.githubusercontent.com/67451029/96650805-5902cb00-12e8-11eb-92bf-ce27013b4bcc.png)

tgregg commented on tests/test_res_diff.py:
```diff
@@ -0,0 +1,47 @@
+TEMP_FILE = os.path.join(TEST_FILE_PATH, 'temp')
+
+
+def pytest_generate_tests(metafunc):
+    metafunc.parametrize('file', TEST_FILES)
```
I'm new to this parameterization technique. Can you paste a screen snip of what it looks like to run these tests in PyCharm?

cheqianh replied:
We have 43 test cases and they all passed.

![Screen Shot 2020-10-21 at 3 22 21 PM](https://user-images.githubusercontent.com/67451029/96793420-3a1c3b80-13b1-11eb-9f60-4ee795dc6b47.png)

tgregg replied:
Okay, thanks.

tgregg commented on amazon/iontest/ion_test_driver.py:
```diff
@@ -741,13 +749,397 @@ def parse_implementations(descriptions, output_root):
+def get_name_for_write(location):
+    location_array = location.split("/")
+    # implementation's name
+    impl = location_array[-4]
```
Should the location_array's size be validated so that a helpful error may be raised if it's malformed?

cheqianh replied:
Sure, in case people modify the result file manually. 

The path after `results/` are created by ion-test-driver. I added code to check if the file structure are valid for children under the dir `'results'`.

tgregg commented on amazon/iontest/ion_test_driver.py:
```diff
@@ -741,13 +749,397 @@ def parse_implementations(descriptions, output_root):
+    return disagree_list
+
+
+def find_disagree_lists_for_write(failures_list, impl):
```
The logic in this method looks like it's performed twice: once for the first impl and one for the second impl. I recommend refactoring the common logic into method(s) and invoking those method(s) multiple times, to cut down on code duplication.

cheqianh replied:
ion_test_driver 827-842.

tgregg commented on amazon/iontest/ion_test_driver.py:
```diff
@@ -741,13 +749,397 @@ def parse_implementations(descriptions, output_root):
+    return disagree_lists
+
+
+def analyze_list(first_list, second_list, first_impl, second_impl):
```
This looks like another method that could be refactored into a smaller method that is called twice.

cheqianh replied:
Ion-test-driver.py 845-856.

tgregg commented on amazon/iontest/ion_test_driver.py:
```diff
@@ -741,13 +749,397 @@ def parse_implementations(descriptions, output_root):
+    return no_more_agree_list, start_agree_list
+
+
+def analyze_lists(first_lists, second_lists):
```
Same thing here.

cheqianh replied:
Ion-test-driver.py 859 - 876.

tgregg commented on amazon/iontest/ion_test_driver.py:
```diff
@@ -741,13 +749,397 @@ def parse_implementations(descriptions, output_root):
+        temp_dir = os.path.join(dir_name, 'temp')
+        if os.path.exists(temp_dir):
+            shutil.rmtree(temp_dir)
+        _, stderr = Popen((TOOL_DEPENDENCIES['git'], 'clone', '--recursive', des_list[1], temp_dir),
+                          stderr=PIPE, shell=COMMAND_SHELL).communicate()
+        os.chdir(temp_dir)
+        _, stderr = Popen((TOOL_DEPENDENCIES['git'], 'checkout', des_list[2]),
+                          stderr=PIPE, shell=COMMAND_SHELL).communicate()
```
Should these calls use `check_output`?

cheqianh replied:
Changed.

tgregg commented on tests/res_diff_expect/fail_diff_results.ion:
```diff
@@ -0,0 +1,8 @@
+{
+ test: {
+  result: {
+   message: "Result: two revisions have different results."
```
Should we specify that the new revision fails while the other one succeeded? Are there other cases that cause this?

cheqianh replied:
Different result fields are allowed now. 

One result is PASS and the other one is FAIL will keep going until we found the difference between them and write it into the report.

tgregg commented on tests/res_diff_expect/fail_read_compare_vs_other_errors_2.ion:
```diff
@@ -0,0 +1,46 @@
+      }
+     ]
+    },
+    'ion-java_2': {
+    }
```
It would be helpful to show any errors that ion-java_2 has as well (even those in a different phase), if possible.

cheqianh replied:
We check report phase by phase and end analyzing the file once we found errors.
For this test case that ion-java_2 has a phase three error but the analysis detects errors in phase two (one has read_compare error and the other one not), so analysis for this file end in phase two and we don't have phase three errors included in ion-java_2 report.

The reason to end it immediately once we find an error is that because if one phase has already had an error, the later phase error may not be accurate or caused by earlier phase so we don't spend unnecessary time on analysis. Do you think it's a good strategy?

tgregg replied:
Okay, let's stick with that for now then.

tgregg commented on tests/res_diff_expect/fail_read_disagree.ion:
```diff
@@ -0,0 +1,24 @@
+     "ion-js_1"
+    ]
+   },
+   no_longer_agrees_with: [
+    "ion-c_1"
+   ],
+   now_agrees_with: [
+    "ion-js_1"
+   ]
```
Should these be the other way around? It looks like ion-java_1 agreed with ion-js, but ion-java_2 disagrees with it. And ion-java_1 disagreed with ion-c, but ion-java_2 agrees with it.

cheqianh replied:
Sorry for the confusion, changed the expect file and the logic.

tgregg commented on tests/res_diff_expect/fail_revisions_with_diff_list.ion:
```diff
@@ -0,0 +1,17 @@
+  read_compare: {
+   message: "Read_compare: two revisions agree with each other but have different disagree lists. ",
+   disagree_lists: {
+    'ion-java_1': [
+     "ion-c_1",
+     "ion-js_1"
+    ],
+    'ion-java_2': [
+     "ion-c_1"
+    ]
```
I think this is the case that should be a success and not a failure. ion-java_2's disagree list is a strict subset of of ion-java_1's disagree list, and the two revisions agree with each other. I think we should print this report but not fail.

cheqianh replied:
Sure, 

but in case I misunderstand it, I guess they should have the same disagree list as long as two implementations agree with each other right? 

If impl_1 agrees with impl_2 and impl_1 disagrees with impl_3. then impl_2 has to disagree with impl_3 right?

tgregg replied:
The goal is to have no disagreements between implementations. So if impl_1 disagrees with impl_3, but impl_2 agrees with impl_3 and does not disagree with any additional implementations, then that is a step in the right direction. We should still include the difference in the results, but we should not fail the analysis because of this.

In other words, if impl_2's disagree list is a strict subset of impl_1's disagree list, then that is a positive change.

cheqianh replied:
Got it, changed the logic so it only alerts users when it going the worse direction. This test renamed as `'pass_revisions_with_diff_list.ion'` now.

I also added another test case `'fail_revisions_with_diff_list.ion'` which went the incorrect direction and start disagrees other implementations, which returns non-zero and alert users.

tgregg commented on tests/res_diff_expect/fail_write_compare_vs_other_errors.ion:
```diff
@@ -0,0 +1,19 @@
+  write_error: {
+   message: "Write_error: new commit has different write error(s).",
+   write_error: {
+    'ion-java_1': [
+    ],
+    'ion-java_2': {
+     errors: ErrorReport::[
+      {
+       message: "test"
+      }
```
It would be nice to include ion-java_1's errors too, even if they occur in a different phase.

cheqianh replied:
Similar as the other comment.

tgregg commented on tests/res_diff_tests/pass_fix_read_compare_errors.ion:
```diff
@@ -0,0 +1,22 @@
+             }
+        },
+        'ion-java_2': {
+            result:FAIL
```
Did you mean for this to be `result:PASS`?

cheqianh replied:
I added more test cases to separate this situation.

**This file trying to tes**t: Someone fixes a read_compare error but result is still `FAIL`. For this we should write the diff into the report but not alert users.
**New test case**: Someone fixes a read_compare error so now the new commit's result is `PASS`. show difference but don't alert users.

Did above change for four test cases: `fix read error,` `fix read compare error`, `fix write error` and `fix write compare errors`.

tgregg commented on tests/res_diff_tests/pass_fix_write_compare_errors.ion:
```diff
@@ -0,0 +1,22 @@
+             }
+        },
+        'ion-java_2': {
+            result:FAIL
```
Did you mean for this to be `result:PASS`? If not, we should probably have a file that tests that case.

cheqianh replied:
Added more tests.

tgregg reviewed (commented):
This is coming along really well.

In addition to my comments below, I'd like to see a few more `pass_` tests that cover the case where ion-java_1 has `FAIL` for various reasons, while ion-java_2 has `PASS`. We want to make sure that the results show the difference, but we should expect success since this is an improvement.

cheqianh committed:
Fixed issues and added new tests.

cheqianh commented:
new test cases added:
pass_different_results_1.ion
pass_different_results_1.ion
pass_different_results_1.ion
pass_different_results_1.ion
fail_different_results_1.ion
fail_different_results_1.ion

cheqianh commented:
Thanks for the CR,

### New test cases added:
**pass_different_results_1.ion**: new commit pass but old commit fail with phase one errors.
**pass_different_results_2.ion**: new commit pass but old commit fail with phase two errors.
**pass_different_results_3.ion**: new commit pass but old commit fail with phase three errors.
**pass_different_results_4.ion**: new commit pass but old commit fail with phase four errors.
**fail_different_results_1.ion**: : old commit pass but new commit fail with phase one errors.
**fail_different_results_2.ion**: old commit pass but new commit fail with phase two errors.
**fail_different_results_3.ion**:  old commit pass but new commit fail with phase three errors.
**fail_different_results_4.ion**: old commit pass but new commit fail with phase four errors.
**fail_different_results_failures_1.ion**: old commit pass but new commit fail with failures.
**fail_different_results_failures_2.ion** new commit pass but old commit fail with failures.

cheqianh committed:
Fixed some issues and added some message for better UX.

cheqianh committed:
Fixed typo

tgregg commented on tests/res_diff_expect/fail_different_results_failures_2.ion:
```diff
@@ -0,0 +1,25 @@
+      }
+     ]
+    },
+    'ion-java_2': {
```
Shouldn't this be a `pass_` test? The old revision failed, while the new one did not, and no failure was expected.

cheqianh replied:
Doesn't like Errors, Failures should be equivalence between two revisions because they use its own cli tool to compare the same resources (e.g. read files generated by phase one). 

It's a little tricky between `same failures` and `same disagree list`. we should check if the failures are same first to make sure cli tool behavior same then we extract the disagree list from the failures and analyze the disagree list. Let me know if my understand is incorrect!

tgregg replied:
I don't know if I fully understand your comment, but ion-java_1 had result `FAIL`, while ion-java_2 had result `PASS`. Shouldn't that always cause success since it is a move in the right direction?

cheqianh replied:
Sorry, I forgot to mentioned this earlier that PASS or FAIL is not used for describing behavior good or not. it's just showing if we find difference between implementations.

For error field, the report is showing the errors for an implementation so it's going good direction with less errors. For failures field, the report is showing the difference between implementations so it's slightly different from error field that we need always look into that and extract disagree list.

For example, the old implementation is able to find differences between two implementations so it shows 'FAIL' and writes failures into report while new one's compare command doesn't work and simply shows 'PASS'.

tgregg replied:
Ah, I see. It's because the difference detected is between ion-java_1 and ion-java_2. So this must always be symmetrical, because they would be comparing the same data. Thanks for explaining.

cheqianh committed:
Changed logic for disagree list and added more tests.

cheqianh commented:
Now we have 44 tests and they all passed! 

![Screen Shot 2020-10-22 at 1 54 17 PM](https://user-images.githubusercontent.com/67451029/96928895-17039180-146e-11eb-8261-9fbb3cc2ef86.png)

cheqianh committed:
Fixed indentation issue.

tgregg reviewed (approved):
Very nice! Thanks for going through several rounds.

Can you try this with your ion-java fork to make sure it works as expected in Github Actions? Add the necessary logic to the Actions configurations, then make some small commits that are / aren't expected to trigger a failure due to --results-diff.

cheqianh commented:
Thanks! 

Ok I'll open a new PR related to 'integrating --result-diff options into GitHub Actions.'
